File: libs/ai-kernel/src/handlers/prompts.rs
use axum::{
    Json,
    extract::{Path, State},
};
use chrono::Utc;
use sqlx::{query_as, types::Json as SqlJson};
use uuid::Uuid;

use crate::{
    AppState,
    domain::llm::provider,
    models::prompt_template::{
        CreatePromptTemplateRequest, ListPromptTemplatesResponse, PromptTemplate,
        PromptTemplateRow, PromptVersion, RenderPromptRequest, RenderPromptResponse,
        UpdatePromptTemplateRequest,
    },
};

use super::{ServiceResult, bad_request, db_error, not_found};

async fn load_prompt_row(
    db: &sqlx::PgPool,
    prompt_id: Uuid,
) -> Result<Option<PromptTemplateRow>, sqlx::Error> {
    query_as::<_, PromptTemplateRow>(
        r#"
		SELECT
			id,
			name,
			description,
			category,
			status,
			tags,
			latest_version_number,
			versions,
			created_at,
			updated_at
		FROM ai_prompt_templates
		WHERE id = $1
		"#,
    )
    .bind(prompt_id)
    .fetch_optional(db)
    .await
}

pub async fn list_prompts(
    State(state): State<AppState>,
) -> ServiceResult<ListPromptTemplatesResponse> {
    let rows = query_as::<_, PromptTemplateRow>(
        r#"
		SELECT
			id,
			name,
			description,
			category,
			status,
			tags,
			latest_version_number,
			versions,
			created_at,
			updated_at
		FROM ai_prompt_templates
		ORDER BY updated_at DESC, created_at DESC
		"#,
    )
    .fetch_all(&state.db)
    .await
    .map_err(|cause| db_error(&cause))?;

    Ok(Json(ListPromptTemplatesResponse {
        data: rows.into_iter().map(Into::into).collect(),
    }))
}

pub async fn create_prompt(
    State(state): State<AppState>,
    Json(body): Json<CreatePromptTemplateRequest>,
) -> ServiceResult<PromptTemplate> {
    if body.name.trim().is_empty() || body.content.trim().is_empty() {
        return Err(bad_request("prompt name and content are required"));
    }

    let version = PromptVersion {
        version_number: 1,
        content: body.content.trim().to_string(),
        input_variables: body.input_variables,
        notes: body.notes.unwrap_or_else(|| "Initial version".to_string()),
        created_at: Utc::now(),
        created_by: None,
    };

    let row = query_as::<_, PromptTemplateRow>(
        r#"
		INSERT INTO ai_prompt_templates (
			id,
			name,
			description,
			category,
			status,
			tags,
			latest_version_number,
			versions
		)
		VALUES ($1, $2, $3, $4, 'active', $5, 1, $6)
		RETURNING
			id,
			name,
			description,
			category,
			status,
			tags,
			latest_version_number,
			versions,
			created_at,
			updated_at
		"#,
    )
    .bind(Uuid::now_v7())
    .bind(body.name.trim())
    .bind(body.description)
    .bind(body.category)
    .bind(SqlJson(body.tags))
    .bind(SqlJson(vec![version]))
    .fetch_one(&state.db)
    .await
    .map_err(|cause| db_error(&cause))?;

    Ok(Json(row.into()))
}

pub async fn update_prompt(
    State(state): State<AppState>,
    Path(prompt_id): Path<Uuid>,
    Json(body): Json<UpdatePromptTemplateRequest>,
) -> ServiceResult<PromptTemplate> {
    let Some(current) = load_prompt_row(&state.db, prompt_id)
        .await
        .map_err(|cause| db_error(&cause))?
    else {
        return Err(not_found("prompt template not found"));
    };

    let current_prompt: PromptTemplate = current.clone().into();
    let mut versions = current.versions.0;

    if body.content.is_some() || body.input_variables.is_some() || body.notes.is_some() {
        let base_version = current_prompt.current_version;
        let next_version_number = versions
            .last()
            .map(|version| version.version_number + 1)
            .unwrap_or(1);

        versions.push(PromptVersion {
            version_number: next_version_number,
            content: body
                .content
                .unwrap_or(base_version.content)
                .trim()
                .to_string(),
            input_variables: body.input_variables.unwrap_or(base_version.input_variables),
            notes: body
                .notes
                .unwrap_or_else(|| format!("Version {next_version_number}")),
            created_at: Utc::now(),
            created_by: None,
        });
    }

    let latest_version_number = versions
        .last()
        .map(|version| version.version_number)
        .unwrap_or(current.latest_version_number);

    let row = query_as::<_, PromptTemplateRow>(
        r#"
		UPDATE ai_prompt_templates
		SET name = $2,
			description = $3,
			category = $4,
			status = $5,
			tags = $6,
			latest_version_number = $7,
			versions = $8,
			updated_at = NOW()
		WHERE id = $1
		RETURNING
			id,
			name,
			description,
			category,
			status,
			tags,
			latest_version_number,
			versions,
			created_at,
			updated_at
		"#,
    )
    .bind(prompt_id)
    .bind(body.name.unwrap_or(current_prompt.name))
    .bind(body.description.unwrap_or(current_prompt.description))
    .bind(body.category.unwrap_or(current_prompt.category))
    .bind(body.status.unwrap_or(current_prompt.status))
    .bind(SqlJson(body.tags.unwrap_or(current_prompt.tags)))
    .bind(latest_version_number)
    .bind(SqlJson(versions))
    .fetch_one(&state.db)
    .await
    .map_err(|cause| db_error(&cause))?;

    Ok(Json(row.into()))
}

pub async fn render_prompt(
    State(state): State<AppState>,
    Path(prompt_id): Path<Uuid>,
    Json(body): Json<RenderPromptRequest>,
) -> ServiceResult<RenderPromptResponse> {
    let Some(row) = load_prompt_row(&state.db, prompt_id)
        .await
        .map_err(|cause| db_error(&cause))?
    else {
        return Err(not_found("prompt template not found"));
    };

    let prompt: PromptTemplate = row.into();
    let (rendered_content, missing_variables) = provider::interpolate_template(
        &prompt.current_version.content,
        &body.variables,
        body.strict,
    );

    if body.strict && !missing_variables.is_empty() {
        return Err(bad_request("missing prompt variables"));
    }

    Ok(Json(RenderPromptResponse {
        prompt_id,
        version_number: prompt.current_version.version_number,
        rendered_content,
        missing_variables,
    }))
}

File: libs/ai-kernel-go/handlers/prompts.go
package handlers

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/openfoundry/openfoundry-go/libs/ai-kernel-go/domain/llm"
	"github.com/openfoundry/openfoundry-go/libs/ai-kernel-go/models"
)

// PromptsHandlers is the prompt-templates surface (list / create /
// update / render).
type PromptsHandlers struct {
	Pool *pgxpool.Pool
}

const promptColumns = `id, name, description, category, status, tags,
                       latest_version_number, versions, created_at, updated_at`

// scanPromptRow scans + builds the PromptTemplate, deriving
// current_version from the last entry in versions (or a placeholder
// when versions is empty — matches Rust From impl).
func scanPromptRow(s toolScanner) (models.PromptTemplate, error) {
	var p models.PromptTemplate
	var tagsRaw, versionsRaw []byte
	if err := s.Scan(
		&p.ID, &p.Name, &p.Description, &p.Category, &p.Status,
		&tagsRaw, &p.LatestVersionNumber, &versionsRaw,
		&p.CreatedAt, &p.UpdatedAt,
	); err != nil {
		return p, err
	}
	if len(tagsRaw) > 0 {
		_ = json.Unmarshal(tagsRaw, &p.Tags)
	}
	if p.Tags == nil {
		p.Tags = []string{}
	}
	if len(versionsRaw) > 0 {
		_ = json.Unmarshal(versionsRaw, &p.Versions)
	}
	if p.Versions == nil {
		p.Versions = []models.PromptVersion{}
	}
	if n := len(p.Versions); n > 0 {
		p.CurrentVersion = p.Versions[n-1]
	} else {
		p.CurrentVersion = models.PromptVersion{
			VersionNumber:  p.LatestVersionNumber,
			InputVariables: []string{},
			CreatedAt:      p.CreatedAt,
		}
	}
	return p, nil
}

func (h *PromptsHandlers) loadPromptRow(ctx context.Context, id uuid.UUID) (*models.PromptTemplate, error) {
	row := h.Pool.QueryRow(ctx,
		`SELECT `+promptColumns+` FROM ai_prompt_templates WHERE id = $1`, id)
	p, err := scanPromptRow(row)
	if errors.Is(err, pgx.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	return &p, nil
}

// ListPrompts handles `GET /api/v1/prompts`.
func (h *PromptsHandlers) ListPrompts(w http.ResponseWriter, r *http.Request) {
	rows, err := h.Pool.Query(r.Context(),
		`SELECT `+promptColumns+` FROM ai_prompt_templates
          ORDER BY updated_at DESC, created_at DESC`)
	if err != nil {
		dbError(w, err)
		return
	}
	defer rows.Close()
	out := make([]models.PromptTemplate, 0)
	for rows.Next() {
		p, err := scanPromptRow(rows)
		if err != nil {
			dbError(w, err)
			return
		}
		out = append(out, p)
	}
	writeJSON(w, http.StatusOK, models.ListPromptTemplatesResponse{Data: out})
}

// CreatePrompt handles `POST /api/v1/prompts`. Validates name +
// content. Inserts version 1 with notes defaulting to "Initial version".
func (h *PromptsHandlers) CreatePrompt(w http.ResponseWriter, r *http.Request) {
	var body models.CreatePromptTemplateRequest
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	if strings.TrimSpace(body.Name) == "" || strings.TrimSpace(body.Content) == "" {
		writeError(w, http.StatusBadRequest, "prompt name and content are required")
		return
	}

	description := derefString(body.Description, "")
	category := derefString(body.Category, models.DefaultPromptCategory)
	notes := derefString(body.Notes, "Initial version")

	inputVars := body.InputVariables
	if inputVars == nil {
		inputVars = []string{}
	}
	tags := body.Tags
	if tags == nil {
		tags = []string{}
	}

	version := models.PromptVersion{
		VersionNumber:  1,
		Content:        strings.TrimSpace(body.Content),
		InputVariables: inputVars,
		Notes:          notes,
		CreatedAt:      time.Now().UTC(),
	}

	tagsJSON, _ := json.Marshal(tags)
	versionsJSON, _ := json.Marshal([]models.PromptVersion{version})

	row := h.Pool.QueryRow(r.Context(),
		`INSERT INTO ai_prompt_templates
              (id, name, description, category, status, tags,
               latest_version_number, versions)
            VALUES ($1, $2, $3, $4, 'active', $5, 1, $6)
            RETURNING `+promptColumns,
		uuid.New(), strings.TrimSpace(body.Name), description, category,
		tagsJSON, versionsJSON)
	p, err := scanPromptRow(row)
	if err != nil {
		dbError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, p)
}

// GetPrompt handles `GET /api/v1/prompts/{id}`.
func (h *PromptsHandlers) GetPrompt(w http.ResponseWriter, r *http.Request, promptID uuid.UUID) {
	p, err := h.loadPromptRow(r.Context(), promptID)
	if err != nil {
		dbError(w, err)
		return
	}
	if p == nil {
		writeError(w, http.StatusNotFound, "prompt template not found")
		return
	}
	writeJSON(w, http.StatusOK, p)
}

// UpdatePrompt handles `PATCH /api/v1/prompts/{id}`. When content,
// input_variables, OR notes is supplied, appends a new version
// (latest+1) and bumps latest_version_number; otherwise just
// updates the metadata.
func (h *PromptsHandlers) UpdatePrompt(w http.ResponseWriter, r *http.Request, promptID uuid.UUID) {
	current, err := h.loadPromptRow(r.Context(), promptID)
	if err != nil {
		dbError(w, err)
		return
	}
	if current == nil {
		writeError(w, http.StatusNotFound, "prompt template not found")
		return
	}

	var body models.UpdatePromptTemplateRequest
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}

	versions := append([]models.PromptVersion(nil), current.Versions...)
	if body.Content != nil || body.InputVariables != nil || body.Notes != nil {
		base := current.CurrentVersion
		nextNumber := int32(1)
		if n := len(versions); n > 0 {
			nextNumber = versions[n-1].VersionNumber + 1
		}
		content := base.Content
		if body.Content != nil {
			content = *body.Content
		}
		inputVars := base.InputVariables
		if body.InputVariables != nil {
			inputVars = *body.InputVariables
		}
		var notes string
		if body.Notes != nil {
			notes = *body.Notes
		} else {
			notes = "Version " + itoa32(nextNumber)
		}
		versions = append(versions, models.PromptVersion{
			VersionNumber:  nextNumber,
			Content:        strings.TrimSpace(content),
			InputVariables: inputVars,
			Notes:          notes,
			CreatedAt:      time.Now().UTC(),
		})
	}

	latestVersionNumber := current.LatestVersionNumber
	if n := len(versions); n > 0 {
		latestVersionNumber = versions[n-1].VersionNumber
	}

	name := derefString(body.Name, current.Name)
	desc := derefString(body.Description, current.Description)
	category := derefString(body.Category, current.Category)
	status := derefString(body.Status, current.Status)
	tags := current.Tags
	if body.Tags != nil {
		tags = *body.Tags
	}
	if tags == nil {
		tags = []string{}
	}
	tagsJSON, _ := json.Marshal(tags)
	versionsJSON, _ := json.Marshal(versions)

	row := h.Pool.QueryRow(r.Context(),
		`UPDATE ai_prompt_templates SET
            name = $2, description = $3, category = $4, status = $5,
            tags = $6, latest_version_number = $7, versions = $8,
            updated_at = NOW()
          WHERE id = $1
          RETURNING `+promptColumns,
		promptID, name, desc, category, status, tagsJSON,
		latestVersionNumber, versionsJSON)
	p, err := scanPromptRow(row)
	if err != nil {
		dbError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, p)
}

// RenderPrompt handles `POST /api/v1/prompts/{id}/render`. Loads the
// template's current_version content, runs llm.InterpolateTemplate
// against the request variables, and returns the rendered content +
// missing variable list. In strict mode missing variables → 400.
func (h *PromptsHandlers) RenderPrompt(w http.ResponseWriter, r *http.Request, promptID uuid.UUID) {
	current, err := h.loadPromptRow(r.Context(), promptID)
	if err != nil {
		dbError(w, err)
		return
	}
	if current == nil {
		writeError(w, http.StatusNotFound, "prompt template not found")
		return
	}

	var body models.RenderPromptRequest
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}

	rendered, missing := llm.InterpolateTemplate(current.CurrentVersion.Content, body.Variables, body.Strict)
	if missing == nil {
		missing = []string{}
	}
	if body.Strict && len(missing) > 0 {
		writeError(w, http.StatusBadRequest, "missing prompt variables")
		return
	}

	writeJSON(w, http.StatusOK, models.RenderPromptResponse{
		PromptID:         promptID,
		VersionNumber:    current.CurrentVersion.VersionNumber,
		RenderedContent:  rendered,
		MissingVariables: missing,
	})
}

// itoa32 is a small inline integer-to-string helper to avoid pulling
// strconv just for the version-number formatting.
func itoa32(n int32) string {
	if n == 0 {
		return "0"
	}
	var b [11]byte
	i := len(b)
	negative := false
	x := n
	if x < 0 {
		negative = true
		x = -x
	}
	for x > 0 {
		i--
		b[i] = byte('0' + x%10)
		x /= 10
	}
	if negative {
		i--
		b[i] = '-'
	}
	return string(b[i:])
}
